realm/realm-library/src/main/java/io/realm/internal/Collection.java

```diff
@@ -42,27 +40,158 @@ public void onChange(T observer) {
     }
 
     // Custom Collection iterator. It ensures that we only iterate on a Realm collection that hasn't changed.
-    // TODO: Consider to replace RealmResultsIterator implementation by this since it could be shared by the RealmList.
     public static abstract class Iterator<T> implements java.util.Iterator<T> {
-        private final WeakReference<Collection> collectionWeakReference;
+        Collection iteratorCollection;
+        protected int pos = -1;
 
         public Iterator(Collection collection) {
-            collectionWeakReference = new WeakReference<Collection>(collection);
-            collection.stableIterators.add(new WeakReference<Iterator>(this));
+            this.iteratorCollection = collection;
+
+            if (collection.sharedRealm.isInTransaction()) {
+                detach();
+            } else {
+                iteratorCollection.sharedRealm.addIterator(this);
+            }
         }
 
-        protected void checkRealmIsStable() {
-            Collection collection = collectionWeakReference.get();
-            if (collection != null) {
-                for (WeakReference<Iterator> it : collection.stableIterators) {
-                    if (it.get() == this) {
-                        return;
-                    }
-                }
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean hasNext() {
+            checkValid();
+            return pos + 1 < iteratorCollection.size();
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public T next() {
+            checkValid();
+            pos++;
+            if (pos >= iteratorCollection.size()) {
+                throw new NoSuchElementException("Cannot access index " + pos + " when size is " + iteratorCollection.size() +
+                        ". Remember to check hasNext() before using next().");
             }
-            throw new ConcurrentModificationException(
-                    "No outside changes to a Realm is allowed while iterating a RealmResults." +
-                            " Don't call Realm.refresh() while iterating.");
+            return get(pos);
+        }
+
+        /**
+         * Not supported by Realm collection iterators.
+         *
+         * @throws UnsupportedOperationException
+         */
+        @Deprecated
+        public void remove() {
+            throw new UnsupportedOperationException("remove() is not supported by RealmResults iterators.");
+        }
+
+        void detach() {
+            iteratorCollection = iteratorCollection.createSnapshot();
+        }
+
+        // The iterator become invalid after receiving a remote change notification. In Java, the destruction of
```

Comment:
No, that would be bad. I was mostly thinking of it as an optimization, but if it introduces all kinds of weird behaviour in our implementation it might not be worth it.